File: Code/app/src/main/java/com/bigfriedicecream/recipes/presenters/MainPresenter.java
package com.bigfriedicecream.recipes.presenters;


import android.content.Context;

import com.bigfriedicecream.recipes.interfaces.IMainContract;
import com.bigfriedicecream.recipes.observables.EventsDispatcher;
import com.bigfriedicecream.recipes.observables.RecipeRepository;

import java.util.HashMap;
import java.util.Observable;
import java.util.Observer;

public class MainPresenter implements IMainContract.Presenter, Observer {

    private IMainContract.View view;
    private RecipeRepository recipeRepository;
    private EventsDispatcher eventsDispatcher;

    public MainPresenter(IMainContract.View v) {
        view = v;
        recipeRepository = RecipeRepository.getInstance();
        eventsDispatcher = EventsDispatcher.getInstance();
    }

    public void start(Context c) {
        eventsDispatcher.addObserver(this);
        recipeRepository.load(c);
    }

    public void stop(Context c) {
        eventsDispatcher.deleteObserver(this);
    }

    public void load() {
        view.renderList();
    }

    @Override
    public void update(Observable observable, Object o) {
        HashMap<String, String> hashMap = (HashMap<String, String>)o;
        if (observable instanceof EventsDispatcher) {
            if (hashMap.get("name").equals("NAVIGATION_RECIPE")) {
                String id = hashMap.get("id");
                view.renderRecipe(id);
            }
        }
    }
}

File: Code/app/src/main/java/com/bigfriedicecream/overcooked/presenters/MainPresenter.kt
package com.bigfriedicecream.overcooked.presenters


import android.content.Context

import com.bigfriedicecream.overcooked.interfaces.IMainContract
import com.bigfriedicecream.overcooked.observables.EventsDispatcher
import com.bigfriedicecream.overcooked.observables.RecipeRepository

import java.util.HashMap
import java.util.Observable
import java.util.Observer

class MainPresenter(private val view: IMainContract.View) : IMainContract.Presenter, Observer {
    private val recipeRepository:RecipeRepository = RecipeRepository.instance
    private val eventsDispatcher:EventsDispatcher = EventsDispatcher.instance

    override fun start(c:Context) {
        eventsDispatcher.addObserver(this)
        recipeRepository.load(c)
    }

    override fun stop(c:Context) {
        eventsDispatcher.deleteObserver(this)
    }

    override fun load() {
        view.renderList()
    }

    override fun update(observable: Observable, o: Any) {
        val hashMap:HashMap<String, String> = o as HashMap<String, String>
        if (observable is EventsDispatcher) {
            if (hashMap["name"] == "NAVIGATION_RECIPE") {
                val id:String = hashMap["id"]!!
                view.renderRecipe(id)
            }
        }
    }
}
